### PR TITLE
AdamW operator (Fixing Weight Decay Regularization in Adam)

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Implement a modification of Adam in "Fixing Weight Decay Regularization in Adam" https://arxiv.org/abs/1711.05101.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@sxjscience  @szhengac could you guys help review this PR?

File: python/mxnet/optimizer/optimizer.py [hunk 1]
```diff
@@ -1018,6 +1018,70 @@ class ccSGD(SGD):
...
+    regularization on the weights to the loss: it regularizes weights with large
+    gradients more than L2 regularization would, which was shown to yield better
+    training loss and generalization error in the paper above.
+
+    Updates are applied by::
+
+        rescaled_grad = clip(grad * rescale_grad, clip_gradient)
+        m = beta1 * m + (1 - beta1) * rescaled_grad
+        v = beta2 * v + (1 - beta2) * (rescaled_grad**2)
+        w = w - learning_rate * (m / (sqrt(v) + epsilon) + wd * w)
```
[the remark this answers is not in the document]
Comment (sxjscience) [reply python/mxnet/optimizer/optimizer.py]: I think it's acceptable as long as the `wd` is set correctly.

File: python/mxnet/optimizer/optimizer.py [hunk 2]
```diff
@@ -1018,6 +1018,70 @@ class ccSGD(SGD):
...
+        coef2 = 1. - self.beta2**t
+        lr *= math.sqrt(coef2)/coef1
+
+        kwargs = {'beta1': self.beta1, 'beta2': self.beta2, 'epsilon': self.epsilon,
+                  'rescale_grad': self.rescale_grad}
+        if self.clip_gradient:
+            kwargs['clip_gradient'] = self.clip_gradient
+
+        mean, var = state
+        adamw_update(weight, grad, mean, var, out=weight, lr=lr, wd=wd, **kwargs)
```
Comment (sxjscience) [inline python/mxnet/optimizer/optimizer.py]: Should we set `wd` to something like `wd / self._original_lr`?
